text/0000-layer-metadata-files.md

```diff
@@ -0,0 +1,176 @@
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+## Alternative 1
+
+We could add a prefix to each of the layer metadata files like `layer.`. This would allow us to re-use the cached layers but may lead to awkwardness around naming the layer metadata files - for eg. if a layer is called `layer.build` the metadata file would end up being `layer.layer.build.toml`
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (buildpack arg `$1` )
+│   ├── build.toml
+│   ├── launch.toml
+│   ├── store.toml
+│   ├── layer.<layer>.toml
+│   └── <layer>
+└── config
+    └── metadata.toml
+```
+
+## Alternative 2
+
+We could add a prefix for each of the special metadata files and reserve it for CNB usage. For eg. `launch.toml` could be renamed to `cnb.launch.toml`. This way, we will only be prohibiting layers that have a prefix `cnb.` or a layer named `cnb`. The downsides to this alternative are the fact that the metadata files would not be usable across builds and we would still be prohibiting a class of layer names that may be valid.
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (buildpack arg `$1` )
+│   ├── cnb.build.toml
+│   ├── cnb.launch.toml
+│   ├── cnb.store.toml
+│   ├── <layer>.toml
+│   └── <layer>
+└── config
+    └── metadata.toml
+```
+
+
+## Alternative 3
+
+The top level `<buildpack-workspace>` would house any files and folder that are applicable at a `buildpack` level. Any layer specific files and directory structure would be moved to the `layers` sub-directory inside the `<buildpack-workspace>`.
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (buildpack arg `$1` )
+│   ├── build.toml
+│   ├── launch.toml
+│   ├── store.toml
+│   └── layers
+│       ├── <layer>.toml
+│       └── <layer>
+└── config
+    └── metadata.toml
+```
+
+
+## Alternative 4
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers-dir) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (`CNB_BP_LAYERS_DIR`)
+│   ├── <layer>.toml
+│   └── <layer>
+└── config
+    ├── metadata.toml
+    └── <escaped_buildpack_id> (`CNB_BP_CONFIG_DIR`  )
+        ├── build.toml
+        ├── launch.toml
+        └── store.toml
+```
+
+This structure doesn't break layer re-use and has the same length for path names but the top level directory is called `layers` when it houses more than just layers. And also the `config` directory may clash with the `escaped_buildpack_id`.
+
+## Why is this proposal the best?
+
+Although this is a drastic change, this also open up possibilities in the future to introduce other top level concepts inside the `CNB_BP_CONFIG_DIR` apart from the current ones while still allowing layer names to be free-form and without any restrictions.
```

Comment:
I find this very convincing. I like this alternative because it give us more flexibility to make non-breaking additions to the spec in the future 🙂

Reply:
I totally agree! I will add that I think it's worth making a one drastic change over a small change that we'll need to change again soon.

Reply:
Do we have ideas of things we might eventually want to add under `/output` besides `/layers` and `/config`?